sklearn/cluster/k_means_.py

```diff
@@ -164,19 +164,19 @@ def _tolerance(X, tol):
     return np.mean(variances) * tol
 
 
-def _check_sample_weight(X, sample_weight):
+def _check_normalize_sample_weight(sample_weight, X):
     """Set sample_weight if None, and check for correct dtype"""
-    n_samples = X.shape[0]
-    if sample_weight is None:
-        return np.ones(n_samples, dtype=X.dtype)
-    else:
-        sample_weight = np.asarray(sample_weight)
-        if n_samples != len(sample_weight):
-            raise ValueError("n_samples=%d should be == len(sample_weight)=%d"
-                             % (n_samples, len(sample_weight)))
+
+    sample_weight_was_none = sample_weight is None
+
+    sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+    if not sample_weight_was_none:
```

Comment:
I don't think this `sample_weight_was_none` thing is necessary.
```python
sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
if sample_weight.sum() != 1:
    # normalize the weights to sum up to n_samples
    n_samples = len(sample_weight)
    scale = n_samples / sample_weight.sum()
    sample_weight *= scale
return sample_weight
```
should be enough

Reply:
edited

Reply:
Yeah, but we would be computing one extra sum with the above, also float equality comparison is not that sure. Another solution could be,
```py
sample_weight_validated = _check_sample_weight(sample_weight, X, dtype=X.dtype)
if sample_weight is None:
    # normalize the weights to sum up to n_samples
    scale = n_samples / sample_weight.sum()
    sample_weight_validated *= scale
return sample_weight_validated
```
I can change for it if you prefer. Either way defining one extra temp variable is necessary..

Reply:
You're right, there's no need to compute a sum when not necessary.

> I can change for it if you prefer.

no need